src/instructlab/eval/answers.py

```diff
@@ -0,0 +1,137 @@
+    temperature_config,
+)
+
+
+def reorg_answer_file(answer_file):
+    """Sort by question id and de-duplication"""
+    answers = {}
+    with open(answer_file, "r", encoding="utf-8") as fin:
+        for l in fin:
+            qid = json.loads(l)["question_id"]
+            answers[qid] = l
+
+    qids = sorted(list(answers.keys()))
+    with open(answer_file, "w", encoding="utf-8") as fout:
+        for qid in qids:
+            fout.write(answers[qid])
+
+
+def get_answer(
+    question: dict,
+    model: str,
+    num_choices: int,
+    max_tokens: int,
+    answer_file: str,
+    force_temperature: float,
+):
+    """Answer a question with the model"""
+    assert force_temperature is None or question.get("required_temperature") is None
+    if force_temperature is not None:
+        temperature = force_temperature
+    elif "required_temperature" in question.keys():
+        temperature = question["required_temperature"]
+    elif question["category"] in temperature_config:
+        temperature = temperature_config[question["category"]]
+    else:
+        temperature = 0.7
+
+    choices = []
+    for i in range(num_choices):
+        conv = get_conversation_template(model)
+
+        turns = []
+        for j in range(len(question["turns"])):
+            conv.append_message(conv.roles[0], question["turns"][j])
+            conv.append_message(conv.roles[1], None)
+
+            output = chat_completion_openai(model, conv, temperature, max_tokens)
+
+            conv.update_last_message(output)
+            turns.append(output)
+
+        choices.append({"index": i, "turns": turns})
+
+    # Dump answers
+    ans = {
+        "question_id": question["question_id"],
+        "answer_id": shortuuid.uuid(),
+        "model_id": model,
+        "choices": choices,
+        "tstamp": time.time(),
+    }
+
+    os.makedirs(os.path.dirname(answer_file), exist_ok=True)
+    with open(answer_file, "a", encoding="utf-8") as fout:
+        fout.write(json.dumps(ans) + "\n")
+
+
+def generate_answers(
+    model_name,
+    branch=None,
+    output_dir="eval_output",
+    data_dir=None,
+    question_begin=None,
+    question_end=None,
+    force_temperature=None,
+    num_choices=1,
+    max_tokens=1024,
+    max_workers=1,
+    model_api_base=None,
+    bench_name="mt_bench",
+):
+    """Generate model answers to be judged"""
+    if model_api_base is not None:
+        openai.api_base = model_api_base
+
+    if data_dir is None:
+        data_dir = os.path.join(os.path.dirname(__file__), "data")
+
+    data_base_dir = bench_dir(data_dir, bench_name, branch)
+    output_base_dir = bench_dir(output_dir, bench_name, branch)
+
+    question_file = f"{data_base_dir}/question.jsonl"
+    questions = load_questions(question_file, question_begin, question_end)
+
+    answer_file = f"{output_base_dir}/model_answer/{model_name}.jsonl"
+    if os.path.isfile(answer_file):
+        os.remove(answer_file)
+    # print(f"Output to {answer_file}")
+
```

Comment:
Should uncomment or remove - not opposed to keeping for debugging purposes (can always remove later)

Reply:
Done